Implementation of the MNASNet family of models

1e100 opened the pull request:
As described in https://arxiv.org/pdf/1807.11626.pdf

Training program and the detailed training notes are posted separately, as requested by @fmassa. Imagenet results for depth multiplier 1.0 are as follows:

  * Top-1 = 73.512
  * Top-5 = 91.544

Paper top 1 is 74%. The best result was achieved at epoch 198 out of 200, so it seems likely that the model would benefit from adding more epochs.

Trainer program, checkpoint, training log, and tensorboard files can be found here: https://github.com/1e100/mnasnet_trainer/tree/master

1e100 commented:
MNASNet 0.5 result:

Top1: 66.997% (paper number 67.8%)
Top5: 86.916%

Hyperparameters:

250 epochs
Peak LR = 0.8
Warmup 5 epochs
No weight decay
SGD+Nesterov
Momentum 0.9

I can post the checkpoint, training log, and tensorboard dump.

depthwise commented:
Seems like linter failed on existing models, not on the new files.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,176 @@
+    """
+
+    def __init__(self, num_classes, alpha, dropout=0.2):
+        super().__init__()
```
this is Python3-only.
Can you replace instead by `super(MNASNet, self).__init__()`?

depthwise replied:
Done

1e100 replied:
Done

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,176 @@
+                m.bias.data.zero_()
+
+
+class MNASNet0_5(MNASNet):
```
What about having `mnasnet0_5` etc as functions, instead of classes, in the same way as what we have for the other models?

depthwise replied:
Done

depthwise replied:
Sorry, wrong account again. :-) I'm the same guy as 1e100 though.

1e100 replied:
Done

1e100 commented:
Most of my attention was focused on 0.5 and 1.0 depth multipliers. The former is a good "fast" configuration, the latter is a good "precise" configuration. 

I have also trained the 0.75 once (current best top-1 is 70.31, paper number is 71.5), and 1.3 is still training with 65 more epochs to go, as it takes a lot longer to train. 

Based on my experience with 0.5 and 1.0, I think I just need to find better training hyperparameters and/or train for longer. I'm fundamentally limited by having only 8 consumer-grade GPUs at my disposal.

rwightman commented:
This model looks very promising, excited to try it out. Thanks for the impl. A few comments..

1. The layout of the model changes based on whether dropout is 0. or > 0. This is problematic, not just because people often expect to be able to just swap/reference the classifier by name, as a Linear or 1x1 Conv2d layer, but also because it will prevent models from loading correctly with different settings for dropout. Also, the drop rate isn't passed through to the actual Dropout module.

2. Less important, and more a matter of style and preference. When working with models it's definitely nice to leverage the advantages of PyTorch and build them with class modules as building blocks. Pushing more towards the ol Torch all Sequence style can make code compact but definitely has some drawbacks when debugging, modifying, working with the weights, etc. 

Compare a snippet of the weight names of this model:
```
layers.11.1.layers.4.weight
layers.11.1.layers.4.bias
layers.11.1.layers.6.weight
layers.11.1.layers.7.weight
layers.11.1.layers.7.bias
layers.12.0.layers.0.weight
layers.12.0.layers.1.weight
layers.12.0.layers.1.bias
layers.12.0.layers.3.weight
layers.12.0.layers.4.weight
```
to torchvision ResNet:
```
layer1.2.bn3.weight
layer1.2.bn3.bias
layer2.0.conv1.weight
layer2.0.bn1.weight
layer2.0.bn1.bias
layer2.0.conv2.weight
layer2.0.bn2.weight
layer2.0.bn2.bias
layer2.0.conv3.weight
layer2.0.bn3.weight
layer2.0.bn3.bias
layer2.0.downsample.0.weight
layer2.0.downsample.1.weight
layer2.0.downsample.1.bias
layer2.1.conv1.weight
layer2.1.bn1.weight
layer2.1.bn1.bias
```

1e100 commented:
I don't believe that whether or not dropout is there will prevent the model from loading, as dropout doesn't actually store anything in the checkpoint. In PyTorch dropout probability is defined programmatically. If I take a look at the model dict in the checkpoint I can't see a parameter for dropout in there even when dropout is present in the model. 

The "classifier" module stores the following in the checkpoint:
 - 'classifier.1.weight'
 - 'classifier.1.bias'

Even though dropout has non-zero probability.

rwightman commented:
> I don't believe that whether or not dropout is there will prevent the model from loading, as dropout doesn't actually store anything in the checkpoint. In PyTorch dropout probability is defined programmatically. If I take a look at the model dict in the checkpoint I can't see a parameter for dropout in there even when dropout is present in the model.
> 
> The "classifier" module stores the following in the checkpoint:
> 
> * 'classifier.1.weight'
> * 'classifier.1.bias'
> 
> Even though dropout has non-zero probability.

If you set dropout to 0. and try to load weights created with dropout > 0. you'll get an exception. The extra Sequence container creates another level in the parameter key.

RuntimeError: Error(s) in loading state_dict for MNASNet:
	Missing key(s) in state_dict: "classifier.weight", "classifier.bias". 
	Unexpected key(s) in state_dict: "classifier.1.weight", "classifier.1.bias".

1e100 commented:
I see what you mean, but the model would not work correctly in this case even if loaded, because the checkpoint would be loaded into a model that does not match the checkpoint. I think I'd _prefer_ the model to barf on load in such a case.

So I'm kind of reluctant to address this. @fmassa, what do you think?

1e100 commented:
Looks like PyTorch scales the dropout during training so it can be just omitted during inference. To simplify things I just removed the condition for the dropout. It'll be identity of p is zero anyway.

rwightman commented:
I'd still suggest you leaving the classifier as nn.Linear and pull the dropout out as an optional functional call. It's the most common approach and it's nice not to be astonished. Makes the model easier to work with too. Current checkpoint can be easily remapped with a load/key-map/save.

1e100 commented:
@rwightman PTAL. This is more consistent with other models in torchvision, thanks for the suggestion. I still need to update my published checkpoint, didn't have the time to do it today.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,166 @@
+
+    def forward(self, input):
+        if self.apply_residual:
+            return self.layers.forward(input) + input
```
can you avoid calling `.forward`, and instead dispatch directly to `self.layers(input) + input`?

1e100 replied:
Done

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,166 @@
+        if self.apply_residual:
+            return self.layers.forward(input) + input
+        else:
+            return self.layers.forward(input)
```
same comment here

1e100 replied:
Done

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,166 @@
+_BN_MOMENTUM = 1 - 0.9997
+
+
+class _InvertedResidual(nn.Module):
```
Is this significantly different from https://github.com/pytorch/vision/blob/74b6a750c7b748d03a3c80f29ebf17ca98dcab8b/torchvision/models/mobilenet.py#L14-L40 ?

fmassa replied:
Forget about this comment for now, let's leave each file in `models` to be self-contained

1e100 replied:
Removed the comment, although yes, I do implement my own MobileNet V2 with the IR block you see in my code. :-)

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,166 @@
+        self.num_classes = num_classes
+        self.dropout = dropout
+        depths = _scale_depths([24, 40, 80, 96, 192, 320], alpha)
+        layers = [
```
The more I look at this, the more I have the impression it looks a lot like MobileNet V2 https://github.com/pytorch/vision/blob/74b6a750c7b748d03a3c80f29ebf17ca98dcab8b/torchvision/models/mobilenet.py#L49-L74

Would there be some benefits of making both implementations share a bit more building blocks?

fmassa replied:
Forget about this comment for now, let's leave each file in `models` to be self-contained

rwightman replied:
@fmassa same building blocks as MobileNetV2, FBNet, and ChamNet. I just realized this after doing my own impl of MnasNet yesterday to have the SE variant ('A1' by Google's naming). Then I noticed the comparisons in the FBNet paper, and then the FBNet backbone def in maskrcnn-benchmark (https://github.com/facebookresearch/maskrcnn-benchmark/blob/464b1af167d3dcb513964e62389825ff507609c3/maskrcnn_benchmark/modeling/backbone/fbnet_builder.py)

With maybe a few small tweaks all of the MNasNet architectures can be defined with the above building code, as can all of the other mentioned networks. It'd be nice to have them all together here with the same generic builder :)

1e100 replied:
Speaking of SE, that would not be difficult to add here as a mode, though I never used it myself in practice.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,166 @@
+
+    def forward(self, x):
+        x = self.features(x)
+        x = self.avgpool(x).squeeze()
```
Having a `squeeze()` can lead to surprising results, for example if the batch size is 1, then the batch dimension will be squeezed out as well.

I think in this case it's just better / simpler to do
```python
x = x.mean([2, 3])
```
which will remove the spatial dimension, as expected.

1e100 replied:
Very nice, thanks for the suggestion. Done

1e100 commented:
Trained up another MNASNet 0.5 with the changes (and with more epochs), with substantially better results:
```
loss=1.367, metrics = prec1=67.484, prec5=87.482
```
Paper number for top1 is 67.8. Will upload the checkpoint and training long to the trainer repo.

1e100 commented:
@fmassa is there anything else I need to change? I think I resolved everything, not sure you've seen it yet. PTAL. If there's anything you're not satisfied with, I have the time this week to address it.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+                m.bias.data.zero_()
+
+
+def mnasnet0_5(num_classes):
```
could you add a `pretrained` argument (defaulting to False) as the first argument here, and to the other functions?
Also, do you have the pretrained checkpoints for them?

1e100 replied:
I only have pretrained checkpoints for 0.5 and 1.0. I'm also unsure where I'm supposed to store the checkpoints. Please advise.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+    def _initialize_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
```
can you use `nn.init` methods here and in the next cases? They can be probably replaced by `kaiming_normal_` or `xavier_normal_`.

1e100 replied:
Maybe, maybe not. I'd need to re-train with kaiming_normal_ to know if it's kosher here or not. I think I trained with either of those two (don't recall which) and couldn't get to the accuracy I achieved with this. I'll kick off a run on 0.75 with kaiming_normal_ to see if it's the initialization or something else.

fmassa replied:
It is possible to keep the same initialization hyperparameters parameters while using `kaiming_normal_` / `xavier_normal_`, all you'd need to do is to find the corresponding `a` / `gain` that is needed to match values here.

1e100 replied:
Actually it looks like you're right. The exact example from the documentation looks like it's identical mathematically: `nn.init.kaiming_normal_(w, mode='fan_out', nonlinearity='relu')`. It just didn't occur to me to use `mode='fan_out'`. 

I have replaced the initialization with `kaiming_normal_` and used the `nn.init` module directly, like in mobilenetv2 in this repo.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+        # Equivalent to global avgpool and removing H and W dimensions.
+        x = x.mean([2, 3])
+        if self.dropout > 0.0:
+            x = nn.functional.dropout(x, p=self.dropout, training=self.training,
```
I think it is just less error-prone to just add a `nn.Dropout` in the `classifier`, as a `nn.Sequential`

1e100 replied:
Done.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+
+        self._initialize_weights()
+
+    def features(self, x):
```
Do we need this method? This is not really usual in `torchvision` modules, and we are trying to come up with an unified way of querying intermediate features.

1e100 replied:
This is handy for when you don't want the classifier, such as with YOLO2 for example. You simply take an existing net and bolt a decoder on. But it's of limited use for anything else. I'll remove it.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+        return self.layers(x)
+
+    def forward(self, x):
+        x = self.features(x)
```
Can you just use `layers`, or call the `layers` attribute `features` here?

1e100 replied:
Done, features is gone

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+    def __init__(self, num_classes, alpha, dropout=0.2):
+        super(MNASNet, self).__init__()
+        self.alpha = alpha
+        self.num_classes = num_classes
```
the `num_classes` is not used here, and even though it might be useful in a few cases, the fact that only a handful of models would have it makes things more complicated.

For uniformity, it would be better to remove this, and at a later stage maybe think of a way of unifying some attributes over all models

1e100 replied:
I've removed `self.num_classes`, `self.alpha`, and `self.dropout` and used initializer  parameters directly.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,163 @@
+    """ MNASNet, as described in https://arxiv.org/pdf/1807.11626.pdf.
+    >>> model = MNASNet(1000, 1.0)
+    >>> x = torch.rand(1, 3, 224, 224)
+    >>> y = model.forward(x)
```
can you do `model(x)` and avoid calling `.forward`? This is discouraged in PyTorch, as we lose the hooks attached to the model

1e100 replied:
Done

fmassa commented on torchvision/models/__init__.py:
```diff
@@ -6,3 +6,4 @@
 from .densenet import *
 from .googlenet import *
 from .mobilenet import *
+from .mnasnet import *
```
Can you also add an entry in https://github.com/pytorch/vision/blob/master/docs/source/models.rst ?

1e100 replied:
Done, though I'm not sure I've done it correctly. Please take a look.

1e100 commented:
OK, @fmassa, I think I have addressed all of your suggestions other than initialization, and uploaded a better MNASNet 0.5 checkpoint. 

Regarding initialization, these low-compute nets are more sensitive to normalization, and I'm not actually sure that using the more off-the-shelf stuff will result in the same accuracies. Just to test that out, I'm kicking off a training run on 0.5, with only initializations different. It'll take about 2.5 days. 

Let's get the rest of this ironed out in the meanwhile.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,184 @@
+        raise ValueError(
+            "No checkpoint is available for model type {}".format(model_name))
+    checkpoint_url = _MODEL_URLS[model_name]
+    model.load_state_dict(torch.utils.model_zoo.load_url(checkpoint_url))
```
I think you forgot to use `load_state_dict_from_url` here?

1e100 replied:
Looks like you're right. Fixed.

1e100 commented:
Hi @fmassa. I've addressed your feedback. PTAL

1e100 commented:
Merged master. Good time to merge. :-)

1e100 commented:
@fmassa, is there anything I need to change here?

1e100 commented:
@fmassa, this has been sitting in the queue long enough that I'm considering withdrawing the PR. If this is not going to be merged, please let me know. If I don't hear back anything within a week, I'll close the PR.

fmassa commented:
I'm very sorry for the delay in replying, I was pretty busy with the 0.3 release of torchvision, and some follow-up tasks.

I'm reviewing it today

fmassa commented on docs/source/models.rst:
```diff
@@ -111,6 +114,7 @@ ShuffleNet V2                     30.64           11.68
 MobileNet V2                      28.12           9.71
 ResNeXt-50-32x4d                  22.38           6.30
 ResNeXt-101-32x8d                 20.69           5.47
+MNASNet 1.0                       26.49           8.456
```
Quick question:

I've tried testing the models with the `references/train.py` file, and got
```
 * Acc@1 73.456 Acc@5 91.510
```
which corresponds to `26.54` and `8.49`.

I wonder if there is a difference in our data, or if the model that I downloaded is not the same?

depthwise replied:
I'll verify and report back (and train a better model if needed). Could be something as dumb as the wrong version of Pillow. To aid with investigation, please answer the following questions:
 - Are you using Pillow-SIMD? 
 - Are you compiling it from source? 
 - Are you using libjpeg-turbo8?

depthwise replied:
And also, did you change the number of epochs when training? I think that particular model actually used more epochs.

rwightman replied:
FWIW, I trained this 'b1' variant of mnasnet to 74.66 and the 'a1' SE variant to 75.45. They both took over 400 epochs, both using RMSprop with Google like h-params. Using EMA of the model weights was necessary to match/surpass the papers... 

Models only: https://github.com/rwightman/gen-efficientnet-pytorch#pretrained

Training code w/ models: https://github.com/rwightman/pytorch-image-models

fmassa replied:
@depthwise I don't think that Pillow / the different image processing is the cause.
I've done extensive experiments in the past with multiple models, and they all were fairly insensitive to using PIL / OpenCV. If that is indeed the case, then I'd be surprised, and that could indicate that the model is very fragile to small perturbations.

For completeness, I used Pillow 5.4.1, from pip, using libjpeg-turbo8

@rwightman interesting, and what happens if you do not apply EMA, do you recall what accuracies you get?

rwightman replied:
@fmassa I was topping out around where Dmitry ended up, in the 73-73.9 range over 4 attempts. After adding my EMA impl, I resumed my best training sessions from a little before where the validation error and started to level off, and ran it with EMA for another 60-80 epochs ( think I went back to 380 and ended at 450 but have to check) and ending up getting the results I was looking for. I brought the dropout and weight decay back up to paper level when doing this, I had been lowering them before to squeeze out a few .1%  but the EMA smoothed weights benefit from having a little more variability. I keep the color jitter low, in the google training it looks like it's completely disabled for these models.

1e100 replied:
So to investigate this I wrote a simple eval script, which I pushed to <https://github.com/1e100/mnasnet_trainer/blob/master/eval.py>.

The results with Pillow-SIMD/libjpeg-turbo-8 are as follows:
```
Dev/mnasnet_trainer % ./eval.py
Evaluating pretrained mnasnet1_0
1.0769143749256522 [('prec1', 73.490265), ('prec5', 91.53294)]
Evaluating pretrained mnasnet0_5
1.3720299355229553 [('prec1', 67.59815), ('prec5', 87.51842)]
```
Neither of which matches the published numbers exactly. MNASNet 1.0 is slightly worse than in the doc says. MNASNet 0.5 is slightly better than the checkpoint name would imply (67.598% top1 vs 67.592).

The results with "plain" pillow 6.0.0 from PyPl are as follows:
```
% ./eval.py
Evaluating pretrained mnasnet1_0
1.0772113243536072 [('prec1', 73.46037), ('prec5', 91.52099)]
Evaluating pretrained mnasnet0_5
1.372453243756781 [('prec1', 67.606026), ('prec5', 87.50845)]
```
So the top1 for 1.0 gets a bit worse, and for 0.5 it gets a bit better. I've observed such sensitivity with other "efficient" models in the past. In particular, the resize algorithm (which is different in Pillow SIMD) seems to make a noticeable difference. The effect of this on smaller models is easily measurable. Something as mundane as a different JPEG decoder affects them, and so do the software versions: CUDA/cuDNN, PyTorch, etc - a number of these are different between then and now.

Just to be on the safe side, though, I kicked off another run for 1.0. I should have the results sometime over the weekend, one way or the other.

fmassa replied:
@1e100 if you are running the training again, would you mind using the code from `references/classification`, and maybe only changing the lr step?

For MobileNet V2, I used the following commands, on 8 GPUs
```
--model mobilenet_v2 --epochs 300 --lr 0.045 --wd 0.00004 --lr-step-size 1 --lr-gamma 0.98
```

The most important thing we are trying to do here is to have a simple path for reproducible research, so having a few % worse accuracy (~0.3-0.5 e.g.,) but with a reproducible script available in torchvision would be preferable I'd say.

This way, we can indeed compare apples with apples

1e100 replied:
@fmassa what version of Pillow should I use for this? I use custom-compiled PillowSIMD for my own training runs, but if we want things to be more repeatable, I could use the slower, default install of PillowSIMD.

Here's how I compile it on my machines:
```bash
CC="cc -march=native -mtune=native -O3" pip3 install \
    --force-reinstall --user pillow-simd
```

fmassa replied:
@1e100 sorry, I missed your message.

You can use current version that you have available, don't worry about potentially small differences.

1e100 replied:
Still training. First training run wasn't too good (I tweaked the training regime, and this model is sensitive to that kind of thing), so I'm using that as initialization on the second run. I'll update (and send a PR) when I get a good result.

fmassa commented on torchvision/models/mnasnet.py:
```diff
@@ -0,0 +1,183 @@
+
+_MODEL_URLS = {
+    "mnasnet0_5":
+    "https://github.com/1e100/mnasnet_trainer/releases/download/v0.1/mnasnet0.5_top1_67.592-7c6cb539b9.pth",
```
I have a follow-up commit that updates this

scb-vs5 commented:
Being a freshman to cv, I just can not find the codes about reinforce learning what the paper described（the rewards or the way to update）.
Could you give some tips about which part of the codes  excute the RNN controller?
Thank you so much!

1e100 commented:
@scb-vs5 There's no neural architecture search here, only the model. And only one kind of model (the "B" variant), as well. Unless you have access to immense amounts of spare GPU/TPU capacity like Google does, neural architecture search is not really feasible for you as an individual anyway. In my "back of the napkin" estimation it'd cost no less than $100K to just run the search for the model architecture, and that excludes the time and resources it'd take you to build the infrastructure and get it running reliably. So while I could implement this, it'd be kind of pointless for me to do so, other than to educate. TorchVision, in my view, is more practically-oriented.

scb-vs5 commented:
Thank you for your reply, Then whether it's well recommended for individual to start “NAS” with weights sharing or gradient policy ，which could speed up the train and practically-oriented？Like ENas or darts?

fmassa commented:
@scb-vs5 I would recommend asking such questions in the PyTorch forums http://discuss.pytorch.org/

azamatkhid commented:
Hello! Recently, I tried to train the mnasnet1_0 from scratch on cifar10 dataset (with some small augmentations), but the loss behavior was very weird, as well the training accuracy does not exceed 10%.

fmassa commented:
@azamatkhid I would recommend checking how the model was trained on ImageNet following the instructions from https://github.com/1e100/mnasnet_trainer

MNasNet might have a few training tricks that are not really standard.

1e100 commented:
@azamatkhid Training procedure there is tuned for large datasets like ImageNet. You may need to e.g. reduce momentum for batchnorm, or even leave it at its default value for a small dataset like CIFAR. FWIW, I've been using this model quite a lot lately for object detection on smaller datasets with great success. I use BN momentum of 0.01 on such datasets (or 0.99 in TensorFlow terms). Try this, see what happens.

Beyond this, please share your training code on GitHub. It's difficult to see what's wrong with so little information. Could be a million different things.

azamatkhid commented:
@fmassa and @1e100 thanks for your comments. 
This is the repository with the code:
https://github.com/azamatkhid/mnasnet-pytorch
I will try to re-train with BN momentum 0.1, and let you know guys if it successful. 
In addition, I was wondering why there is no mnasnet-a1 official implementation on pytorch?
